[CIAPP-5098] Add TeamCity integration to public docs

andrea-mosk opened the pull request:
<!-- *Note: Please remember to review the Datadog Documentation [Contribution Guidelines](https://github.com/DataDog/documentation/blob/master/CONTRIBUTING.md) if you have not yet done so.* -->

### What does this PR do?
<!-- A brief description of the change being made with this pull request.-->
Add TeamCity integration to public docs

### Motivation
<!-- What inspired you to submit this pull request?-->
We are releasing TeamCity as a new integration

<!-- ### Preview -->
<!-- Assuming you are a Datadog employee and named your branch `<yourname>/<description>`, a preview build will run and links to the preview output will be auto-generated and posted in the PR comments. The links will 404 until the preview build is finished running -->

### Additional Notes
<!-- Anything else we should know when reviewing?-->

---

### Reviewer checklist
- [x] Review the changed files.
- [ ] Review the URLs listed in the [Preview](#preview) section.
- [x] Check images for PII
- [x] Review any mentions of "Contact Datadog support" for internal support documentation.

andrea-mosk commented on content/en/continuous_integration/pipelines/teamcity.md:
```diff
@@ -0,0 +1,95 @@
+<div class="alert alert-info"><strong>Note:</strong> The git author email is used for
+<a href="https://www.datadoghq.com/pricing/?product=ci-visibility#ci-visibility" target="_blank">billing</a> purposes,
+therefore there might be cost implications when either "UserId" or "Author Name" are used as TeamCity username styles.
+Reach out to us if you have any questions about your use case.</div>
```
How can I say "Reach out to us" without incurring in these warnings? Contact Datadog support?

andrea-mosk commented on content/en/continuous_integration/pipelines/teamcity.md:
```diff
@@ -0,0 +1,95 @@
+
+- **Queue time**: View amount of time pipeline jobs sit in the queue before processing
+
+- **Pipeline failure reasons**: Identify pipeline failure reasons from error messages
```
Same as above :/

andrea-mosk commented on content/en/continuous_integration/pipelines/teamcity.md:
```diff
@@ -0,0 +1,95 @@
+
+## Configure the Datadog integration
+
+The integration between [TeamCity][1] and Datadog CI Visibility is provided through a TeamCity plugin.
```
Mmm I prefer "provided through" here

AdrianLC commented on content/en/continuous_integration/pipelines/teamcity.md:
```diff
@@ -0,0 +1,95 @@
+
+- **Partial retries**: View partially retried pipeline executions
+
+- **Queue time**: View amount of time pipeline jobs sit in the queue before processing
```
I would wait for editorial review first. Otherwise you might have to update all of them twice

kassenq replied:
Ah just saw this comment - I think we should be consistent across providers, don't have strong preference for individual wording so will defer to you and/or docs team reviewer

andrea-mosk commented on content/en/continuous_integration/pipelines/teamcity.md:
```diff
@@ -0,0 +1,95 @@
+composite build as their last build are ignored by the plugin.
+3. The following configuration parameters need to be present for the TeamCity projects:
+   * `datadog.ci.api.key`: having as value your [Datadog API Key][2].
+   * `datadog.ci.site`: having as value {{< region-param key="dd_site" code="true" >}}.
```
Not sure I get your question, but the webhooks intake path is https://webhook-intake.{{DD_SITE}}/api/v2/webhook, where DD_SITE can be one of:

- datadoghq.com (default)
- datadoghq.eu
- us3.datadoghq.com
- us5.datadoghq.com

The Plugin just compose the path based on the DD_SITE provided in the parameter. Here, the value can be one of those basically.

AdrianLC replied:
ok that's great I was thinking the templating would give us only us1/us3/eu1 etc

alxch- commented on content/en/continuous_integration/pipelines/teamcity.md:
```diff
@@ -0,0 +1,95 @@
+To activate the integration:
+1. Download the [Datadog CI Integration plugin][5] on the TeamCity server by going to
+Administration -> Plugins -> Browse Plugin Repository.
+2. For the integration to work properly, a [TeamCity composite build][6] must be added as the last build
```
Maybe there's a screenshot you can provide with exactly what people should input there?

andrea-mosk replied:
For the composite build? Unfortunately the dependencies they need to set really depend on how they structured their build chain :/ The only screenshot I can provide is how to create a composite build, but it's basically a "check" when creating a build, to define that it will be composite

alxch- replied:
understood, I think this makes sense to add it anyway and maybe show what it looks like (a random pipeline with a composition build in the end). Sorry if it does not make sense, feel free to ignore

andrea-mosk replied:
I've added a sample pipeline as an image, looks fine to me. WDYT?

alxch- commented on content/en/continuous_integration/pipelines/teamcity.md:
```diff
@@ -0,0 +1,95 @@
+accessed from the TeamCity Server by going to "Administration -> Diagnostic -> Server Logs". If the plugin is not behaving
+correctly, these logs can be checked to get additional context on what might be wrong.
+
+## Git user information
```
I think we should give a recommendation on what people should do here, before we tell them what does not work

andrea-mosk replied:
Makes sense, let me add that part here

andrea-mosk replied:
Changed!

andrea-mosk commented on content/en/continuous_integration/pipelines/teamcity.md:
```diff
@@ -0,0 +1,95 @@
+
+The [Pipelines][3] and [Pipeline Executions][4] pages populate with data after the pipelines finish.
+
+**Note**: The Pipelines page shows data for only the default branch of each repository.
```
Mmm I see that the only related thing that that page says is that the pipelines list won't display data if the default branch is incorrect, which is already implicitly said in the sentence, no?

alxch- replied:
It also explains how to fix it

andrea-mosk replied:
Fair, I've linked the term

kassenq commented:
Nit - for consistency w/ other [provider documentation](https://docs.datadoghq.com/continuous_integration/pipelines/gitlab/?tab=gitlabcom) can we change, on the Queue time line, "View amount of time pipeline jobs sit in the queue before processing" to "View amount of time pipeline jobs wait in the queue before processing?" I don’t have a strong preference for either "sit" or "wait", just want it to be consistent. Otherwise LGTM

andrea-mosk commented:
> Nit - for consistency w/ other [provider documentation](https://docs.datadoghq.com/continuous_integration/pipelines/gitlab/?tab=gitlabcom) can we change, on the Queue time line, "View amount of time pipeline jobs sit in the queue before processing" to "View amount of time pipeline jobs wait in the queue before processing?" I don’t have a strong preference for either "sit" or "wait", just want it to be consistent. Otherwise LGTM

@kassenq I was sure that I copied the queue sentence from another provider, and indeed we use "sit" in Buildkite https://docs.datadoghq.com/continuous_integration/pipelines/buildkite/ 😂 . So I guess we are not consistent at the moment (although this is something negligible).

kassenq commented:
> 

Ah okay! I wasn't aware. Let's merge these docs first and I can make a PR afterward to make the wording consistent

maycmlee commented:
Created [DOCS-4700](https://datadoghq.atlassian.net/browse/DOCS-4700) for docs review.

buraizu reviewed (changes requested):
Hi team, thanks for the PR and the attention paid to the documentation here. I've left a review based on our [style guide](https://github.com/DataDog/documentation/blob/master/CONTRIBUTING.md), and also added my input to some of the active conversations.
Is it possible to change the screenshot so that it does not depict `master` as the default branch? This is due to our guidelines around [inclusive language](https://github.com/DataDog/documentation/blob/master/CONTRIBUTING.md#inclusive-language).
Thanks again, and let me know any questions or feedback.

andrea-mosk commented on content/en/continuous_integration/pipelines/teamcity.md:
```diff
@@ -0,0 +1,104 @@
+The integration between [TeamCity][1] and Datadog CI Visibility is provided through a TeamCity plugin.
+To activate the integration:
+1. Download the [Datadog CI Integration plugin][5] on the TeamCity server by going to
+Administration -> Plugins -> Browse Plugin Repository.
```
Done!

andrea-mosk commented on content/en/continuous_integration/pipelines/teamcity.md:
```diff
@@ -0,0 +1,104 @@
+of the build chain (this step can be skipped if there is already one). Namely, this build must have
+a dependency on the current last build of the chain and no other builds depending on it. Build chains not having a
+composite build as their last build are ignored by the plugin. Below is an example of an expected build chain,
+in which "Aggregating Results" is the last composite build:
```
Changed in this way! Only difference is that I used "build chain" instead of "buildchain".

andrea-mosk commented on content/en/continuous_integration/pipelines/teamcity.md:
```diff
@@ -0,0 +1,104 @@
+
+## Git user information
+
+The plugin retrieves the Git author/committer name and email based on the [TeamCity username style][7].
```
It's fine to remove the committer part, changed

andrea-mosk commented on content/en/continuous_integration/pipelines/teamcity.md:
```diff
@@ -0,0 +1,104 @@
+   value `true` to the Root Project. More information on defining configuration parameters
+   are available in the [TeamCity Project Hierarchy][9] docs.
+4. To enable the plugin, click on "Enable uploaded plugins" in the Administration -> Plugins page.
+Alternatively, restart the TeamCity server.
```
Done!

andrea-mosk commented on content/en/continuous_integration/pipelines/teamcity.md:
```diff
@@ -0,0 +1,104 @@
+
+## Compatibility
+
+- **Supported TeamCity versions**: 2021.2 or higher
```
Changed

andrea-mosk commented on content/en/continuous_integration/pipelines/teamcity.md:
```diff
@@ -0,0 +1,104 @@
+
+## Visualize pipeline data in Datadog
+
+The [Pipelines][3] and [Pipeline Executions][4] pages populate with data after the pipelines finish.
```
Changed

andrea-mosk commented on content/en/continuous_integration/pipelines/teamcity.md:
```diff
@@ -0,0 +1,104 @@
+
+The [Pipelines][3] and [Pipeline Executions][4] pages populate with data after the pipelines finish.
+
+**Note**: The Pipelines page shows data for only the [default branch][12] of each repository.
```
Thanks, changed

andrea-mosk commented on content/en/continuous_integration/pipelines/teamcity.md:
```diff
@@ -0,0 +1,104 @@
+
+All the logs generated by the Datadog CI Integration plugin are stored inside the `teamcity-server.log` file and can be
+accessed from the TeamCity Server by going to "Administration -> Diagnostic -> Server Logs". If the plugin is not behaving
+correctly, these logs can be checked to get additional context on what might be wrong.
```
Changed

andrea-mosk commented on content/en/continuous_integration/pipelines/teamcity.md:
```diff
@@ -0,0 +1,104 @@
+## Git user information
+
+The plugin retrieves the Git author/committer name and email based on the [TeamCity username style][7].
+In this regard, we highly recommend to use either "Author Name and Email" or "Author Email" username styles, as they
+provide information about the user email. When one of the other username styles is used ("UserId" or "Author Name"),
+the plugin automatically generates an email for the user by adding `@Teamcity` at the end of the username.
+For example, the plugin generates `john.doe@Teamcity` as the Git author email when the "UserId" username
+style is used and the git author username is `john.doe`. The username style is defined for [VCS Roots][11],
+and can be modified in the VCS Root settings.
```
Changed, thanks!

andrea-mosk commented on content/en/continuous_integration/pipelines/teamcity.md:
```diff
@@ -0,0 +1,104 @@
+and can be modified in the VCS Root settings.
+
+<div class="alert alert-info"><strong>Note:</strong> The git author email is used for
+<a href="https://www.datadoghq.com/pricing/?product=ci-visibility#ci-visibility" target="_blank">billing</a> purposes,
+therefore there might be cost implications when username styles not providing email ("UserId" or "Author Name") are used.
+Reach out to us if you have any questions about your use case.</div>
```
It's inside an html tag, so I needed to use <strong></strong> to make it bold, but done!

andrea-mosk commented on content/en/continuous_integration/pipelines/teamcity.md:
```diff
@@ -0,0 +1,104 @@
+composite build as their last build are ignored by the plugin. Below is an example of an expected build chain,
+in which "Aggregating Results" is the last composite build:
+{{< img src="ci/teamcity_build_chain.png" alt="TeamCity build chain with composite build at the end" style="width:90%;">}}
+The final composite build should be properly configured in terms of version control settings. It should have
+the VCS Root attached and the [VCS Trigger][13] configured.
```
Yeah, looks good

andrea-mosk commented on content/en/continuous_integration/pipelines/teamcity.md:
```diff
@@ -0,0 +1,104 @@
+{{< img src="ci/teamcity_build_chain.png" alt="TeamCity build chain with composite build at the end" style="width:90%;">}}
+The final composite build should be properly configured in terms of version control settings. It should have
+the VCS Root attached and the [VCS Trigger][13] configured.
+3. The following configuration parameters need to be present for the TeamCity projects:
+   * `datadog.ci.api.key`: having as value your [Datadog API Key][2].
+   * `datadog.ci.site`: having as value {{< region-param key="dd_site" code="true" >}}.
+   * `datadog.ci.enabled`: having as value `true` (`false`
+   can be used to disable the plugin for a specific project).
```
Changed!

andrea-mosk commented on content/en/continuous_integration/pipelines/teamcity.md:
```diff
@@ -0,0 +1,104 @@
+   * `datadog.ci.enabled`: having as value `true` (`false`
+   can be used to disable the plugin for a specific project).
+
+   These configuration parameters can be added either to TeamCity subprojects
+   or to the [TeamCity Root Project][10]. When added to the Root project, they are propagated
+   to all the subprojects. For example, to enable the plugin for all projects, `datadog.ci.enabled` can be added with
+   value `true` to the Root Project. More information on defining configuration parameters
+   are available in the [TeamCity Project Hierarchy][9] docs.
```
Changed, thanks!

andrea-mosk commented:
Thanks for the feedback, @buraizu! I have addressed all the feedback in the last 2 commits and I have also changed the screenshot to not show the `master` branch anymore (I used `test_branch` instead). Feel free to take a look at the changes :) I will take a look at the page preview later in the day or tomorrow morning! 

> Is Datadog support the correct team to answer this question, as it pertains to billing? Should this not be directed to a Customer Success Manager? If so, it looks like this is usually done by linking to mailto:success@datadoghq.com.

I checked this internally with the team, and they should reach out to the support actually

andrea-mosk commented:
👍 Confirming that everything looks good on the preview from my side.

buraizu commented on content/en/continuous_integration/pipelines/teamcity.md:
```diff
@@ -0,0 +1,104 @@
+
+**Note**: The Pipelines page only shows data for the [default branch][12] of each repository.
+
+## Troubleshooting
```
Would it make sense to move the troubleshooting section to the end of the doc, just before the `Further reading` section? This is the format used in most other integration docs.

andrea-mosk replied:
Yeah makes sense! I've moved it

buraizu commented on content/en/continuous_integration/pipelines/teamcity.md:
```diff
@@ -0,0 +1,104 @@
+the plugin generates `john.doe@Teamcity` as the Git author email. The username style is defined for [VCS Roots][11],
+and can be modified in the VCS Root settings.
+
+<div class="alert alert-info"><strong>Note:</strong> The git author email is used for
```
```suggestion
<div class="alert alert-danger"><strong>Note:</strong> The Git author email is used for
```

buraizu replied:
Optional suggestion to update this to be red and more eye-catching because of the potential billing implications.

andrea-mosk replied:
Makes sense to me, I'll double check that visually is still ok

buraizu commented on content/en/continuous_integration/pipelines/teamcity.md:
```diff
@@ -0,0 +1,104 @@
+<div class="alert alert-info"><strong>Note:</strong> The git author email is used for
+<a href="https://www.datadoghq.com/pricing/?product=ci-visibility#ci-visibility" target="_blank">billing</a> purposes,
+therefore there might be cost implications when username styles not providing email
+(<strong>UserId</strong> or <strong>Author Name</strong>) are used. Reach out to us if you have any questions about your use case.
```
```suggestion
(<strong>UserId</strong> or <strong>Author Name</strong>) are used. <a href="https://docs.datadoghq.com/help/" target="_blank">Reach out to the Datadog support team</a> if you have any questions about your use case.
```

buraizu replied:
Thanks for confirming that support is the correct team here!

buraizu reviewed (approved):
Thanks for implementing those changes. I noticed a few minor issues and posted a question about moving the troubleshooting section, but otherwise this LGTM.
